File: computing_methods/SystemOfLinearEquations/src/JacobiMethod.java
import javafx.util.Pair;

public class JacobiMethod implements SolvingMethod {

    public final int iterationsLimit;

    public JacobiMethod(int iterationsLimit) {
        this.iterationsLimit = iterationsLimit;
    }
    @Override
    public Pair<Vector, Integer> solve(LinearEquationsSystem system, double epsilon) {
        Matrix D = Matrix.byDiagonal(system.getA().getDiagonal());
        Matrix dReversed = D.reversed();
        if (dReversed == null) return null;
        Matrix B = Matrix.unit(system.getA().getHeight()).subtract(dReversed.multiply(system.getA()));
        Vector g = dReversed.multiply(system.getB());
        Vector x = Vector.zero(system.getB().getDimensions());
        Vector nextX = x;
        int iterations = 0;
        double q = B.norm();
        do {
            x = nextX;
            nextX = B.multiply(x).add(g);
            iterations++;
            if (nextX.hasNaN() || iterations > iterationsLimit)
                return null;
        } while (!x.equals(nextX, epsilon * (1 - q)));
        return new Pair<>(nextX, iterations);
    }
}

File: computing_methods/CourseWork/src/JacobiMethod.scala
class JacobiMethod(iterationLimit: Int) extends LinearSystemsSolvingMethod {
  override def solve(s: LinearEquationSystem, epsilon: Double): (Vector, Int) = {
    val dReversed = Matrix.byDiagonal(s.getA.getDiagonal).reversed
    if (dReversed == null) return null
    val B = Matrix.unit(s.getA.getHeight) - dReversed * s.getA
    val g = dReversed * s.getB
    val q = B.norm
    var x = Vector.zero(s.getB.getDimension)
    var nextX = x
    var iterations = 0
    do {
      x = nextX
      nextX = B * x + g
      iterations = iterations + 1
      if (nextX.hasNan || iterations > iterationLimit) return null
      //} while (!x.equals(nextX, epsilon * (1 - q)))
    } while ((x - nextX).norm > (1 - q) / q * epsilon)
    (nextX, iterations)
  }
}
